### PR TITLE
added promotions for SymTri and Tri

### LINKED ISSUE #982: Missing promotions in Tridiagonal and SymTridiagonal constructors
This is a suboptimal experience:
```
julia> using LinearAlgebra

julia> SymTridiagonal([1,2,3],[4,5.0])
ERROR: MethodError: no method matching SymTridiagonal(::Vector{Int64}, ::Vector{Float64})
```
Should be as easy as adding a constructor that does promotion?

### LINKED ISSUE #48505: Missing promotions in Tridiagonal and SymTridiagonal constructors
This is a suboptimal experience:
```
julia> using LinearAlgebra

julia> SymTridiagonal([1,2,3],[4,5.0])
ERROR: MethodError: no method matching SymTridiagonal(::Vector{Int64}, ::Vector{Float64})
```
Should be as easy as adding a constructor that does promotion?

### PR DESCRIPTION
**Pull request by ArunS-tack**
closes JuliaLang/LinearAlgebra.jl#982.

### CODE REVIEW

**Comment by inkydragon** [comment]
```
LinearAlgebra/tridiag                            (4) |         failed at 2023-02-05T06:16:15.101
Test Failed at /Users/julia/.julia/scratchspaces/a66863c6-20e8-4ff4-8a62-49f30b1f605e/agent-cache/default-macmini-aarch64-2.0/build/default-macmini-aarch64-2-0/julialang/julia-master/julia-ccae09c640/share/julia/stdlib/v1.10/LinearAlgebra/test/tridiag.jl:92
  Expression: SymTridiagonal(d, GenericArray(dl))
    Expected: MethodError
      Thrown: StackOverflowError
Test Failed at /Users/julia/.julia/scratchspaces/a66863c6-20e8-4ff4-8a62-49f30b1f605e/agent-cache/default-macmini-aarch64-2.0/build/default-macmini-aarch64-2-0/julialang/julia-master/julia-ccae09c640/share/julia/stdlib/v1.10/LinearAlgebra/test/tridiag.jl:93
  Expression: SymTridiagonal(GenericArray(d), dl)
    Expected: MethodError
      Thrown: StackOverflowError
```
`StackOverflowError` implies that there may be infinite recursion in the code.

**Comment by ArunSanganal** [comment]
Looks like this is good to go.

File: stdlib/LinearAlgebra/src/tridiag.jl [hunk 1]
```diff
@@ -73,6 +73,10 @@ function SymTridiagonal{T}(dv::AbstractVector, ev::AbstractVector) where {T}
     SymTridiagonal(convert(AbstractVector{T}, dv)::AbstractVector{T},
                    convert(AbstractVector{T}, ev)::AbstractVector{T})
 end
+function SymTridiagonal(d::Vector{T}, e::Vector{S}) where {T,S}
+    TS = promote_type(T,S)
+    return SymTridiagonal{TS,Vector{TS}}(d, e)
+end
```
Comment (dkarrasch) [inline stdlib/LinearAlgebra/src/tridiag.jl:79]:
```suggestion
SymTridiagonal(d::AbstractVector{T}, e::AbstractVector{S) where {T,S} =
    SymTridiagonal{promote_type(T, S)}(d, e)
```

**Comment by LilithHafner** [comment]
Test failures look real

File: stdlib/LinearAlgebra/src/tridiag.jl [hunk 2]
```diff
@@ -73,6 +73,8 @@ function SymTridiagonal{T}(dv::AbstractVector, ev::AbstractVector) where {T}
     SymTridiagonal(convert(AbstractVector{T}, dv)::AbstractVector{T},
                    convert(AbstractVector{T}, ev)::AbstractVector{T})
 end
+SymTridiagonal(d::AbstractVector{T}, e::AbstractVector{S) where {T,S} =
```
Comment (dkarrasch) [inline stdlib/LinearAlgebra/src/tridiag.jl:76]:
```suggestion
SymTridiagonal(d::AbstractVector{T}, e::AbstractVector{S}) where {T,S} =
```

File: stdlib/LinearAlgebra/src/tridiag.jl [hunk 3]
```diff
@@ -513,12 +515,16 @@ julia> Tridiagonal(dl, d, du)
...
+    Tridiagonal{promote_type(T, S, U, V)}(dl, d, du, du2)
+Tridiagonal{T}(dl::AbstractVector, d::AbstractVector, du::AbstractVector) where {T} =
     Tridiagonal(map(x->convert(AbstractVector{T}, x), (dl, d, du))...)
-end
-function Tridiagonal{T}(dl::AbstractVector, d::AbstractVector, du::AbstractVector, du2::AbstractVector) where {T}
+Tridiagonal{T}(dl::AbstractVector, d::AbstractVector, du::AbstractVector, du2::AbstractVector) where {T} =
     Tridiagonal(map(x->convert(AbstractVector{T}, x), (dl, d, du, du2))...)
-end
+Tridiagonal{T,V}(A::Tridiagonal) where {T,V<:AbstractVector{T}} =
+    Tridiagonal{T,V}(A.dl, A.d, A.du)
```
Comment (dkarrasch) [inline stdlib/LinearAlgebra/src/tridiag.jl:527]:
```suggestion
```

File: stdlib/LinearAlgebra/src/tridiag.jl [hunk 4]
```diff
@@ -73,6 +73,8 @@ function SymTridiagonal{T}(dv::AbstractVector, ev::AbstractVector) where {T}
     SymTridiagonal(convert(AbstractVector{T}, dv)::AbstractVector{T},
```
Comment (dkarrasch) [inline stdlib/LinearAlgebra/src/tridiag.jl:73]:
```suggestion
    SymTridiagonal{T}(convert(AbstractVector{T}, dv)::AbstractVector{T},
```

File: stdlib/LinearAlgebra/src/tridiag.jl [hunk 5]
```diff
@@ -73,6 +73,8 @@ function SymTridiagonal{T}(dv::AbstractVector, ev::AbstractVector) where {T}
...
                    convert(AbstractVector{T}, ev)::AbstractVector{T})
```
Comment (dkarrasch) [inline stdlib/LinearAlgebra/src/tridiag.jl:74]:
```suggestion
    d = convert(AbstractVector{T}, dv)::AbstractVector{T}
    e = convert(AbstractVector{T}, ev)::AbstractVector{T}
    typeof(d) == typeof(e) ? SymTridiagonal{T}(d, e) : throw(ArgumentError("diagonal vectors needed to be promotable to same type"))
```

**Pull request merged**